SP1L0 Integration

mitchellc5 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Integration of SP1L0 as well as addition of new CU1124 into etherCAT topology. 
<!--- Describe your changes in detail -->
Addition of CU1124 to L0S15 which fans out etherCAT into the FEE for MR1L0, MR2L0, and the newly installed SP1L0:DCCM. 

SP1L0:DCCM replaces SP1L0:KMONO which previously ran on LFE-Motion. SP1L0:DCCM is a five-axes double channel crystal monochromator installed in the FEE on L0S09. 

## Motivation and Context
The SP1L0:DCCM is an optic device and was naturally fit to be integrated to lfe-optics as it's position will affect beam delivery to multiple hutches. It will be used by the NEH as well as the FEH who will use the monochromator to help with energy calibration which was previously accomplished using the XPP LODCM (removed for L2HE installation).

## How Has This Been Tested?
Build 2024.35 loaded. I tested SP1L0 stages within NC online and had all SP1L0 b=nEnableMode to ALWAYS. I tested each Bragg axes, Th1 and Th2 at 0.1 degree/s and found no issues with motion or triggering limits during EPS checkout or with the stage. 

I tested Txd and Tyd (diagnostics) at 0.2mm/s and found no issues triggering the limits during EPS checkout or with the stage. There was issues with the encoders reading far off values once encoders were connected. After investigating further, I un-inverted the encoder from the NC and inverted counts in the encoder CoE settings instead. After this, I removed soft limits and jogged to the nearest limit on each axis to ensure the encoder counts matched. Since the encoder was inverted from CoE, I needed to compare the counts by subtracting the encoder readings by 2^64 to see how they matched with FAT and SAT encoder counts. These counts matched up, so I then applied an NC offset to the axes so that the position matched with what would have been expected at the FAT/SAT. These axes are now reading as expected Txd (-12.5mm ->12.5mm) and tyd (48.5mm -> -1mm). 

Tx was tested in the positive direction and ran smoothly, there was errors when moving in the negative direction, but the assumption is that this is a mechanical failure. 

Nick Waters tested motion on MR1L0 and MR2L0 and these devices operated as intended. There was no changes to their code or linking, although MR2L0 was unlinked when the new hardware was scanned. I relinked these axes and they functioned normally. 

## Where Has This Been Documented?
https://confluence.slac.stanford.edu/x/9QyvI

## Screenshots (if appropriate):

## Pre-merge checklist
- [ x] Code works interactively
- [ x] Code contains descriptive comments
- [ x] Test suite passes locally
- [ x] Libraries are set to fixed versions and not ``Always Newest``
- [ x] Code committed with ``pre-commit`` (alternatively ``pre-commit run --all-files``)

nrwslac commented on lcls-plc-lfe-optics/lfe_optics/GVLs/GVL_SP1L0.TcGVL:
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
File looks empty, should delete?

nrwslac commented on lcls-plc-lfe-optics/lfe_optics/GVLs/GVL_SP1L0_Constants.TcGVL:
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
File looks empty, should delete?

mitchellc5 commented:
@nrwslac can you take a look again and review this code. Thank you

nrwslac commented on lcls-plc-lfe-optics/lfe_optics/POUs/Main.TcPOU:
```diff
@@ -347,19 +347,253 @@ VAR
+        pv: DCCM:MONO:SOL
+    '}
+    {attribute 'TcLinkTo' :='TIIB[EL3054_SP1L0_AI]^AI Standard Channel 1^Value'}
+    wAI AT %I*:INT;
```
This should be UINT and n<Name>.

https://confluence.slac.stanford.edu/display/PCDS/ECS+TwinCAT+Standards+and+Style+Guide

Checkout data type prefixes.

nrwslac commented on lcls-plc-lfe-optics/lfe_optics/POUs/Main.TcPOU:
```diff
@@ -347,19 +347,253 @@ VAR
+
+
+
+    // Absolute Encoders for when axes are running in open loop
```
What's the deal with all these left over encoder value variables? Are these left over from the vender? I think we can remove them?

mitchellc5 replied:
Yeah these were left over from the vendor. I had deleted them earlier, but they came back when I reverted the commit and forgot to go back to delete it.

nrwslac commented on lcls-plc-lfe-optics/lfe_optics/POUs/Main.TcPOU:
```diff
@@ -347,19 +347,253 @@ VAR
+
+    bInit : BOOL:=TRUE;
+
+    // Motion Parameters
```
What are these for? They look unused.

mitchellc5 replied:
Same here, they were used by the vendor to make the logic loop. I can remove

mitchellc5 replied:
test also looks unused. I will delete 
![image](https://github.com/user-attachments/assets/0d2efa1d-3781-44b2-9b43-1d4f7ac05c2a)

nrwslac commented on lcls-plc-lfe-optics/lfe_optics/POUs/Main.TcPOU:
```diff
@@ -546,8 +780,73 @@ nEncRefPitchM2L0 := LREAL_TO_UDINT(ABS(fEncRefPitchM2L0_urad) * fEncLeverArm_mm)
+    M13.bEnable:=TRUE;
+    M13.bPowerSelf :=TRUE;
+    M13.nBrakeMode := ENUM_StageBrakeMode.NO_BRAKE;
+    M13.nEnableMode := ENUM_StageEnableMode.ALWAYS;
```
These are already set in the declaration. You can delete these lines.

nrwslac commented on lcls-plc-lfe-optics/lfe_optics/POUs/Main.TcPOU:
```diff
@@ -546,8 +780,73 @@ nEncRefPitchM2L0 := LREAL_TO_UDINT(ABS(fEncRefPitchM2L0_urad) * fEncLeverArm_mm)
+    //Bragg 1
+
+    M13.bEnable:=TRUE;
+    M13.bPowerSelf :=TRUE;
```
This is set in declaration, can delete.

nrwslac commented on lcls-plc-lfe-optics/lfe_optics/POUs/Main.TcPOU:
```diff
@@ -546,8 +780,73 @@ nEncRefPitchM2L0 := LREAL_TO_UDINT(ABS(fEncRefPitchM2L0_urad) * fEncLeverArm_mm)
+
+    M13.bEnable:=TRUE;
+    M13.bPowerSelf :=TRUE;
+    M13.nBrakeMode := ENUM_StageBrakeMode.NO_BRAKE;
```
I've never set a non braked axis explicitly to NO_BRAKE. @NSLentz Should we start doing this?

mitchellc5 replied:
fyi, this was from vendor code.

nrwslac replied:
Gotcha, yea we can delete this. I looked at the implementation briefly. I don't think we need to do any explicit `NO_BRAKE` assigning mostly because we don't have a brake linked anywhere. Although, I was surprised to see brake is the default.

mitchellc5 commented:
I deleted the duplicated code to what's found in the declaration. I also deleted some of the old code from the vendor. I'll have to wait for another PAMM, I think in a couple of weeks to redownload. Should I still commit and push even though it's not the code on the PLC? @nrwslac

mitchellc5 commented:
@nrwslac can you give this one more review?

mitchellc5 commented:
I ran pre-commit 3 times but still failed the pre-commit check. Hopefully it's not an issue

nrwslac reviewed (approved):
There's a few tasks to complete to really tighten this PR up, but they're also part of an overall effort to modernize the format of this project. There are for example variables without device specific naming in main, which could become confusing, but eventually we will move them to their device specific program block and the variables won't need explicit naming.
